detekt-rules/src/main/kotlin/detektive/javacompat/AnnotatePublicApisWithJvmOverloads.kt

```diff
@@ -0,0 +1,91 @@
+package detektive.javacompat
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import io.gitlab.arturbosch.detekt.api.internal.RequiresTypeResolution
+import org.jetbrains.kotlin.descriptors.CallableMemberDescriptor
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+import org.jetbrains.kotlin.descriptors.Visibilities
+import org.jetbrains.kotlin.descriptors.effectiveVisibility
+import org.jetbrains.kotlin.psi.KtAnnotationEntry
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtFunction
+import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.KtPrimaryConstructor
+import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.isPublishedApi
+import kotlin.reflect.KClass
+
+@RequiresTypeResolution
+class AnnotatePublicApisWithJvmOverloads(config: Config) : Rule(config) {
+
+  override val issue = Issue(
+    javaClass.simpleName,
+    Severity.Defect,
+    "Public functions and constructors with default values should be annotated with @JvmOverloads",
+    Debt.FIVE_MINS
+  )
+
+  override fun visitNamedFunction(function: KtNamedFunction) {
+    checkElement(function, ElementType.FUNCTION)
+    super.visitNamedFunction(function)
+  }
+
+  override fun visitPrimaryConstructor(constructor: KtPrimaryConstructor) {
+    checkElement(constructor, ElementType.CONSTRUCTOR)
+    super.visitPrimaryConstructor(constructor)
+  }
+
+  private fun checkElement(element: KtFunction, elementType: ElementType) {
```

Comment:
This is where the main change in this PR is. A custom detekt rule implementation, which find public functions and constructors with one or more default arguments which are missing `@JvmOverloads`

Reply:
I've included this rule within misk just for convenience.  It should really be moved into a separate repository given this is a generic rule. That'd also simplify some of build configs which are making exceptions for this sub-project.